File: leetcode077.py
class Solution:
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        res = []
        self.comb(res,[],n,k,1)
        return res


    def comb(self,res, pre, n,k,start):
        if k == 0:
            res.append(list(pre))
        elif start<=n:
            pre.append(start)
            self.comb(res, pre, n, k-1, start+1)
            pre.pop()
            self.comb(res, pre, n, k, start+1)

if __name__ == "__main__":
    s = Solution()
    print(s.combine(4, 2))

File: leetcodec++/leetcode077.cpp
//
// Created by 黄敏辉 on 2019-07-23.
//

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> comb;
        helper(res, comb, k, n, 1);
        return res;
    }

    void helper(vector<vector<int>>& res, vector<int>& comb, int k, int n, int begin){
        if (comb.size() == k) {
            res.push_back(comb);
            return;
        }

        for (int i = begin; i <= n; i++){
            comb.push_back(i);
            helper(res, comb, k, n, i+1);
            comb.pop_back();
        }
    }
};